### PR TITLE
:sparkles: Add provider support to contract to set provider 

### LINKED ISSUE #4580: eth.Contract objects are never garbage collected
Looking at the code here:
https://github.com/ChainSafe/web3.js/blob/a1c7d71973ec17f9287fbea8939e64a80e589fc6/packages/web3-eth/src/index.js#L332

It looks like all eth.Contract objects reference each other in a chain of setProvider calls. This means eth.Contract objects can not be garbage collected and will just cause a slow memory leak if you have a code that creates them over time. 

Here is an example from a heapdump that shows how setProvider references nest:
image

Also, I suspect if you create lots of eth.Contract objects (maybe 100k, maybe 1million), setProvider call will probably just cause a stack overflow error.

### PR DESCRIPTION
**Pull request by nazarhussain**
## Description

Add support to set providers for multiple contracts. 

Fixes #4580

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network.
- [x] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

**Comment by nazarhussain** [comment]
Please read this thread while reviewing this PR. https://github.com/ChainSafe/web3.js/issues/4580#issuecomment-1086854311

**Review by avkos — APPROVED** [review]

File: packages/web3/src/index.ts [hunk 1]
```diff
@@ -61,6 +61,9 @@ export class Web3 extends Web3Context<EthExecutionAPI> {
 		};
 	};
 
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private readonly _contracts: Contract<any>[] = [];
```
Comment (luu-alex) [inline packages/web3/src/index.ts]: do we ever need to remove contracts?

**Review by luu-alex — APPROVED** [review]

File: packages/web3/src/index.ts [hunk 2]
```diff
@@ -70,6 +73,29 @@ export class Web3 extends Web3Context<EthExecutionAPI> {
...
+			}
+
+			const contract = self.use(Contract, jsonInterface, address, options);
+
+			self._contracts.push(contract);
+
+			return contract;
+		}
+
+		ContractBuilder.setProvider = (_provider: SupportedProviders<EthExecutionAPI>) => {
```
Comment (jdevcs) [inline packages/web3/src/index.ts]: currently contract is using universal provider coming via `Web3Context` > `Web3RequestManager` why do we need to explicitly cascade change after once it set at start.

Comment (nazarhussain) [reply packages/web3/src/index.ts]: To preserve the behaviour of 1.x that once the provider is changed on `Contract` static level, provider must be changed on all instances of any contract ever created in script life cycle.

File: packages/web3/src/index.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/web3/src/index.ts]: As the instance of the `Web3` object will be garbage collected, these contracts will be removed.

**Pull request merged**